File: how-to-print-file-trees-on-the-command-line/snippet.py
import os

def build_tree(entity, indent, is_head, is_tail, result):
	# determine if entity is a directory or file
	files = []
	if os.path.isdir(entity):
		files = os.listdir(entity)

	# keep track of line prefixes
	entity_prefix = '└── ' if is_tail else '├── '
	content_prefix = '    ' if is_tail else '|   '
	if is_head:
		entity_prefix = content_prefix = ''

	# add entity to the output
	result += indent + entity_prefix + os.path.basename(entity) + '\n'

	# if entity is a directory, recurse through its contents
	for index in range(len(files) - 1):
		result = build_tree(os.path.join(entity, files[index]), indent + content_prefix, False, False, result)
	if len(files) > 0:
		result = build_tree(os.path.join(entity, files[-1]), indent + content_prefix, False, True, result)
	return result

def generate_file_tree(entity):
	tree = build_tree(entity, '', True, True, '')
	print(tree)

File: how-to-print-file-trees-on-the-command-line/snippet.js
const fs = require('fs')
const path = require('path')
const util = require('util')
const readdirPromise = util.promisify(fs.readdir)
const statPromise = util.promisify(fs.stat)

const buildTree = async (entity, indent, isHead, isTail, result) => {
    // determine if entity is a directory or file
    let stats = await statPromise(entity)
    let files = []
    if (stats.isDirectory()) {
        files = await readdirPromise(entity)
    }

    // keep track of line prefixes
    let entityPrefix = isTail === true ? '└── ' : '├── '
    let contentPrefix = isTail === true ? '    ' : '|   '
    if (isHead === true) {
        entityPrefix = ''
        contentPrefix = ''
    }

    // add entity to the output
    result += indent + entityPrefix + path.basename(entity) + '\n'

    // if entity is a directory, recurse through its contents
    for (let index = 0; index < files.length - 1; index++) {
        result = await buildTree(path.join(entity, files[index]), indent + contentPrefix, false, false, result)
    }
    if (files.length > 0) {
        result = await buildTree(path.join(entity, files[files.length - 1]), indent + contentPrefix, false, true, result)
    }
    return result
}

const generateFileTree = async (entity) => {
    let tree = await buildTree(entity, '', true, true, '')
    console.log(tree)
}
